### PR TITLE
feat: Update word embedding format to simple context-only with on-the…

### PR DESCRIPTION
**Pull request by tnorlund**
# Pull Request

## 📋 Description

This PR updates the word embedding format in `receipt_chroma` to use a simple context-only format that supports on-the-fly embedding. The new format replaces the XML-style format with a space-separated format that is easier to construct and enables embedding words without full receipt context.

### Key Changes

1. **New Embedding Format**:
   - Replaced XML-style format (`<TARGET>word</TARGET> <POS>position</POS>`) with simple space-separated format (`left_words word right_words`)
   - Configurable context size (default: 2 words on each side, up from 1)
   - Multiple `<EDGE>` tags (one per missing word position) for better edge detection
   - Removed position tags (`<POS>`) as context provides spatial information

2. **On-the-Fly Embedding Support**:
   - Added `format_word_for_embedding()` helper function for embedding words without full receipt context
   - Enables agents to embed words on-the-fly for similarity search
   - Supports hypothetical queries and partial context scenarios

3. **Improved Neighbor Detection**:
   - Updated to find neighbors based on horizontal position (x-coordinate) regardless of line
   - Better captures context from words that are far left/right on other lines
   - Useful for receipts with multi-line items or vertically aligned columns

4. **ChromaDB File Locking Fixes**:
   - Improved garbage collection and file handle release in `chroma_client.py`
   - Enhanced snapshot validation with proper cleanup
   - Addresses SQLite file locking issues (issue #5868)

5. **Metadata Updates**:
   - Renamed `validated_labels` to `valid_labels` for consistency
   - Added `left_context` and `right_context` metadata fields

### Files Changed

**Package (`receipt_chroma/`):**
- `receipt_chroma/data/chroma_client.py` - File locking fixes
- `receipt_chroma/embedding/delta/word_delta.py` - Format updates
- `receipt_chroma/embedding/formatting/word_format.py` - Core format changes
- `receipt_chroma/embedding/metadata/word_metadata.py` - Metadata field rename
- `receipt_chroma/s3/snapshot.py` - File locking fixes
- `receipt_chroma/tests/unit/test_word_format.py` - Comprehensive test updates

**Infrastructure:**
- `infra/chromadb_compaction/` - Updated to use new format
- `infra/embedding_step_functions/` - Updated embedding handlers
- `infra/upload_images/container_ocr/handler/embedding_processor.py` - Format integration

**Documentation:**
- `docs/WORD_EMBEDDING_FORMAT_UPDATE.md` - Comprehensive format documentation
- `docs/AGENT_ON_THE_FLY_EMBEDDING.md` - Agent integration guide

### Example Format

**Old Format:**
```
<TARGET>Total</TARGET> <POS>top-right</POS> <CONTEXT>Subtotal Tax</CONTEXT>
```

**New Format (context_size=2):**
```
Subtotal Items Total Tax Discount
```

**Edge Cases:**
- At left edge: `<EDGE> <EDGE> Total Tax Discount`
- At right edge: `Items Subtotal Total <EDGE> <EDGE>`
- Both edges: `<EDGE> <EDGE> Total <EDGE> <EDGE>`

## 🔄 Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [x] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation update
- [x] 🐛 Bug fix (non-breaking change which fixes an issue) - ChromaDB file locking

**Note**: This is a breaking change for new embeddings. Existing stored embeddings will continue to work for queries, but new embeddings will use the new format.

## 🧪 Testing

- [x] Tests pass locally
- [x] Added tests for new functionality
- [x] Updated existing tests if needed
- [x] Manual testing completed

### Test Coverage

- Updated `test_word_format.py` with comprehensive tests for:
  - New format generation
  - Edge case handling (left/right/both edges)
  - Context size variations
  - Neighbor detection across lines
  - On-the-fly embedding format

## 📚 Documentation

- [x] Documentation updated
- [x] Comments added for complex logic
- [ ] README updated if needed

### Documentation Added

1. **WORD_EMBEDDING_FORMAT_UPDATE.md**: Complete guide to the new format including:
   - Format structure and examples
   - Migration notes
   - Usage examples
   - Implementation details

2. **AGENT_ON_THE_FLY_EMBEDDING.md**: Guide for agent integration including:
   - Use cases
   - Example agent tools
   - Integration points
   - Implementation status

### Cursor Bot Review

- [ ] Waiting for Cursor bot analysis
- [ ] Cursor findings addressed
- [ ] No critical issues found

### AI Review (Cursor)

- [ ] Cursor bot findings addressed (if any)

## ✅ Checklist

- [x] My code follows this project's style guidelines
- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

### Migration Strategy

1. **Existing Embeddings**: Will continue to work for queries but won't match new embeddings
2. **New Embeddings**: Will use the new format going forward
3. **Regeneration**: Consider regenerating embeddings for better consistency (optional)

### Deployment Steps

1. Deploy `receipt_chroma` package update
2. Deploy infrastructure changes (embedding step functions, compaction handlers)
3. Monitor embedding generation to ensure new format is being used
4. Verify ChromaDB file locking improvements in production

### Rollback Plan

If issues arise:
1. Revert this PR
2. Existing embeddings will continue to work
3. No data loss expected (format change only affects new embeddings)

### Performance Impact

- **Positive**: Simpler format may reduce embedding API costs slightly
- **Positive**: Better context capture (2 words vs 1) may improve search quality
- **Neutral**: No significant performance changes expected

### Breaking Changes

- ⚠️ **New embeddings will use different format** - This is intentional and part of the improvement
- ✅ **Query compatibility maintained** - Old queries will still work
- ✅ **Backward compatible** - Existing stored embeddings remain functional

## 📷 Screenshots

N/A - Code changes only

---

### Related Issues

- Addresses ChromaDB file locking issues (#5868)
- Enables on-the-fly embedding for agents
- Improves embedding context quality

### Future Work

- Consider regenerating existing embeddings with new format (optional)
- Add agent tools for on-the-fly embedding (separate PR)
- Update `receipt_label` package to use new format (separate PR)

### References

- See `docs/WORD_EMBEDDING_FORMAT_UPDATE.md` for detailed format documentation
- See `docs/AGENT_ON_THE_FLY_EMBEDDING.md` for agent integration guide

---

**Note**: This PR will be automatically reviewed by Cursor bot. Please address any findings before requesting human review.

## Summary by CodeRabbit

* **New Features**
  * On-the-fly word embedding with context-aware formatting and configurable context_size
  * Automatic metadata enhancement that fills missing merchant details after ingestion

* **Bug Fixes**
  * Improved resource cleanup to reduce file-handle retention and related errors

* **Documentation**
  * Added comprehensive guides for on-the-fly embedding and unified embedding format

* **Chores**
  * Renamed stored label metadata key from "validated_labels" to "valid_labels"

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: infra/embedding_step_functions/simple_lambdas/find_unembedded/handler.py [hunk 1]
```diff
@@ -4,21 +4,110 @@
...
+    line_receipt_dict: dict[str, dict[int, list[ReceiptLine]]],
+) -> list[dict]:
+    """
+    Serialize ReceiptLines into per-receipt NDJSON files.
+    """
+    results: list[dict] = []
+    for image_id, receipts in line_receipt_dict.items():
+        for receipt_id, lines in receipts.items():
+            # Serialize each line as JSON
+            ndjson_lines = [json.dumps(dict(line)) for line in lines]
```
[the remark this answers is not in the document]
Comment (tnorlund) [reply infra/embedding_step_functions/simple_lambdas/find_unembedded/handler.py]: This is not true.
